src/easier68k/assembler/assembler.py

```diff
@@ -37,10 +46,93 @@ def parse(text: str):  # should return a list file and errors/warnings eventuall
+            parsed = parse_literal(contents)
+            # TODO: Max/min memory location?
+            current_memory_location = int.from_bytes(parsed, 'big')
+            continue
+
+        if has_label(stripped):
+            label = get_label(stripped)
+            label_addresses[label] = current_memory_location
+            to_return.define_symbol(label, current_memory_location)
+
+        # We don't know this opcode, there's no module for it
+        if opcode.lower() not in opcodes.__all__:
+            issues.append(('Opcode {} not known, but continuing and dropping it.', 'ERROR'))
+            continue
+
+        op_module = sys.modules['easier68k.core.opcodes.{}'.format(opcode.lower())]
+        op_class = getattr(op_module, opcode.capitalize())
+        length, issues = op_class.get_word_length(opcode, contents)
+
+        current_memory_location += length * 2
+
+    # --- PART 3: actually create the list file ---
+    for line_index, stripped in for_line_stripped_comments(text):
+        opcode = get_opcode(stripped).upper()
+
+        # Equates have already been processed, skip them
+        # (this idea could be expanded for more preprocessor directives)
+        if opcode == 'EQU':
+            continue
+
+        contents = strip_opcode(stripped)
+
+        # Replace all substitutions in the current line with their corresponding values
+        for equate in equates.items():
+            contents = contents.replace(equate[0], equate[1])
+
+        # Replace all memory labels with their proper values (that's the difference in this step)
+        for label in label_addresses.items():
+            contents = contents.replace(label[0], '$' + hex(label[1])[2:])
+
+        if opcode == 'ORG':  # This will shift our current memory location, it's a special case
+            parsed = parse_literal(contents)
+            new_memory_location = int.from_bytes(parsed, 'big')
+            assert 0 <= new_memory_location < 16777216, 'ORG address must be between 0 and 2^24!'
+            current_memory_location = new_memory_location
+            continue
+
+        # We don't know this opcode, there's no module for it
+        if opcode.lower() not in opcodes.__all__:
+            issues.append(('Opcode {} not known, but continuing and dropping it.', 'ERROR'))
+            continue
+
+        # Get the module of this opcode
+        op_module = sys.modules['easier68k.core.opcodes.{}'.format(opcode.lower())]
+
+        # Get the class of this opcode from inside the module
+        op_class = getattr(op_module, opcode.capitalize())
+
+        # Get the actual constructed opcode
+        data, issues = op_class.from_str(opcode, contents)
+
+        # Write the data to the list file
+        to_return.insert_data(current_memory_location, str(binascii.hexlify(data.assemble()))[2:-1])
+
+        # Increment our memory counter
+        current_memory_location += length * 2
+
```

Comment:
I don't think you actually return: to_return anywhere :)

Also I see a test for this, but does it actually check the algorithm? Or is that for another PR?

Reply:
You're right, I didn't have a return yet because I had been testing with console output: that's added now. And I haven't made a formal test yet, mainly because it's hard to test without having all the commands implemented yet, so yeah, a formal test is probably for another PR (for now I'm just testing based on what I'm seeing as output).